controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -88,64 +95,67 @@ public SegmentHelper(final ConnectionFactory clientCF, HostControllerStore hostS
         return Controller.NodeUri.newBuilder().setEndpoint(host.getIpAddr()).setPort(host.getPort()).build();
     }
 
+    private void closeConnection(Reply reply, RawClient client) {
+        log.info("Closing connection as a result of receiving: {}", reply);
+        RawClient c;
+        synchronized (lock) {
+            c = client;
+            client = null;
+        }
+        if (c != null) {
+            try {
+                c.close();
+            } catch (Exception e) {
+                log.warn("Exception tearing down connection: ", e);
+            }
+        }
+    }
+
+    private boolean isReplyWanted(Reply reply, Class[] classArray) {
+        for (Class<?> c : classArray) {
+            if (c.isAssignableFrom(reply.getClass())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @SneakyThrows(ConnectionFailedException.class)
+    private Boolean transformReplyToBoolean(Reply reply, Class[] classArray, RawClient client) {
+        if (isReplyWanted(reply, classArray)) {
+            return true;
+        }
+
+        closeConnection(reply, client);
+        if (reply instanceof WireCommands.NoSuchSegment) {
+            throw new NoSuchSegmentException(reply.toString());
+        } else if (reply instanceof WireCommands.WrongHost) {
+            throw new ConnectionFailedException(reply.toString());
+        } else {
+            throw new ConnectionFailedException("Unexpected reply of " + reply + " when expecting a "
```

Comment:
The semantics of this function are strange. It either returns `true` of throws. If it never returns false, why have a return value at all?